### ISSUE TITLE
start sonarr service with command line within monit 

### ISSUE THREAD

**Issue by fredguy** [SynoCommunity/spksrc#2321]
This is a repost in a new topics as recomended by mod.

Hi guys, 
Wondering if you could share with me the following. 
Sonarr tends to crash from time to time. So i'm triggering an alert with monit. 
Best would be to restrat the service when necessary, but i couldn't figure out the proper start command line. 
/usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe => failed
/var/packages/nzbdrone/scripts/start-stop-status start => failed.

I guess i'm missing some ENV details but can't figure them out. Any help appreciated. 
I'm on DSM5 (not planing to migrate jst yet coz Zarafa is running)
Cheers.

**Comment by Dr-Bean** [comment]
`/var/packages/nzbdrone/scripts/start-stop-status start` should work.

Otherwise, this isn't nearly enough information. Again, take a look at the guidelines when opening an issue: https://github.com/SynoCommunity/spksrc/blob/master/CONTRIBUTING.md#issue-content.

Monit has a log in Package Center. What does that say? What does the Monit configuration file for Sonarr look like? What does Monit's interface say about the monitored service? Does the package not start, or does it take so long that Monit stops monitoring (as Sonarr can take quite a while to start)

**Comment by fredguy** [comment]
yup thought so. Was hopping something was missing in that "/var/packages/nzbdrone/scripts/start-stop-status start"
So let's dive a bit more into the details : 
Before even hoping to get it to work in mmonit, i was looking to get it started from the said command line.

When sonar is working : 
```
  PID USER      PRI  NI  VIRT   RES   SHR S CPU% MEM%   TIME+  Command

 6255 nzbdrone   30  10 2278M  102M  2796 S  1.4  5.2 41:32.32 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe

 6987 nzbdrone   30  10 2278M  102M  2796 S  1.4  5.2  6:45.38 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
26180 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:00.26 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
25485 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:00.17 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
26343 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:00.23 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 6263 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:04.24 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 6271 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:09.90 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 6989 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:07.93 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 6992 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  5:21.88 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 6993 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  8:48.12 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 6994 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2 10:03.02 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 7003 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:02.16 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 7354 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:03.58 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
 7991 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:01.81 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
26268 nzbdrone   30  10 2278M  102M  2796 S  0.0  5.2  0:00.35 /usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
```
/var/packages/nzbdrone/scripts/start-stop-status stop stops the process as expected. 

then /var/packages/nzbdrone/scripts/start-stop-status start i get the : 
Starting Sonarr ...
And nothing.... No process shows up on the console. Even 10 minutes later.
this would be the content of start script (at least the beginning) 
```
#!/bin/sh

# Package
PACKAGE="nzbdrone"
DNAME="Sonarr"

# Others
INSTALL_DIR="/usr/local/${PACKAGE}"
PATH="${INSTALL_DIR}/bin:${PATH}"
USER="${PACKAGE}"
USER_HOME="$(eval echo ~$USER)"
PID_FILE="${USER_HOME}/.config/NzbDrone/nzbdrone.pid"
INSTALL_LOG="${INSTALL_DIR}/var/install.log"
MONO_PATH="/usr/local/mono/bin"
MONO="${MONO_PATH}/mono"
SONARR="${INSTALL_DIR}/share/NzbDrone/NzbDrone.exe"
COMMAND="env PATH=${MONO_PATH}:${PATH} LD_LIBRARY_PATH=${INSTALL_DIR}/lib ${MONO} -- --debug ${SONARR}"

start_daemon ()
{
    start-stop-daemon -c ${USER} -S -q -b -N 10 -x ${COMMAND} > /dev/null
    sleep 2
}

stop_daemon ()
{
    start-stop-daemon -K -q -u ${USER} -p ${PID_FILE}
    wait_for_status 1 20 || start-stop-daemon -K -s 9 -q -p ${PID_FILE}
}
```

So running the following command 
/usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe
gives me : 
```
[Info] Bootstrap: Starting Sonarr - /volume1/@appstore/nzbdrone/share/NzbDrone/NzbDrone.exe - Version 2.0.0.4146
[Info] MigrationLogger: *** Migrating data source=/root/.config/NzbDrone/nzbdrone.db;cache size=-10485760;datetimekind=Utc;journal mode=Wal;pooling=True;version=3 ***

[Fatal] ConsoleApp: EPIC FAIL!

[v2.0.0.4146] System.TypeInitializationException: The type initializer for 'System.Data.SQLite.SQLiteFactory' threw an exception. ---> System.DllNotFoundException: libsqlite3.so.0

  at (wrapper managed-to-native) System.Data.SQLite.UnsafeNativeMethods:sqlite3_config_none (System.Data.SQLite.SQLiteConfigOpsEnum)

  at System.Data.SQLite.SQLite3.StaticIsInitialized () <0x41de2d40 + 0x00053> in <filename unknown>:0

  at System.Data.SQLite.SQLiteLog.Initialize () <0x41de2a90 + 0x0001f> in <filename unknown>:0

  at System.Data.SQLite.SQLiteFactory..cctor () <0x41de0f80 + 0x00057> in <filename unknown>:0

  --- End of inner exception stack trace ---

  at NzbDrone.Core.Datastore.Migration.Framework.MigrationDbFactory.CreateFactory () [0x00000] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Core\Datastore\Migration\Framework\MigrationDbFactory.cs:11

  at FluentMigrator.Runner.Processors.DbFactoryBase.get_Factory () [0x00025] in c:\Community\schambers\fluentmigrator\src\FluentMigrator.Runner\Processors\DbFactoryBase.cs:48

  at FluentMigrator.Runner.Processors.DbFactoryBase.CreateConnection (System.String connectionString) [0x00000] in c:\Community\schambers\fluentmigrator\src\FluentMigrator.Runner\Processors\DbFactoryBase.cs:62

  at NzbDrone.Core.Datastore.Migration.Framework.NzbDroneSqliteProcessorFactory.Create (System.String connectionString, IAnnouncer announcer, IMigrationProcessorOptions options) [0x00006] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Core\Datastore\Migration\Framework\NzbDroneSqliteP
rocessorFactory.cs:14

  at NzbDrone.Core.Datastore.Migration.Framework.MigrationController.Migrate (System.String connectionString, NzbDrone.Core.Datastore.Migration.Framework.MigrationContext migrationContext) [0x00059] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Core\Datastore\Migration\Framework\Migr
ationController.cs:39

  at NzbDrone.Core.Datastore.DbFactory.Create (NzbDrone.Core.Datastore.Migration.Framework.MigrationContext migrationContext) [0x00047] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Core\Datastore\DbFactory.cs:75

  at NzbDrone.Core.Datastore.DbFactory.Create (MigrationType migrationType) [0x00000] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Core\Datastore\DbFactory.cs:49

  at NzbDrone.Core.Datastore.DbFactory.RegisterDatabase (IContainer container) [0x00000] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Core\Datastore\DbFactory.cs:32

  at NzbDrone.Host.Bootstrap.Start (ApplicationModes applicationModes, NzbDrone.Common.EnvironmentInfo.StartupContext startupContext) [0x0003d] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Host\Bootstrap.cs:72

  at NzbDrone.Host.Bootstrap.Start (NzbDrone.Common.EnvironmentInfo.StartupContext startupContext, IUserAlert userAlert, System.Action`1 startCallback) [0x00070] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Host\Bootstrap.cs:39

  at NzbDrone.Console.ConsoleApp.Main (System.String[] args) [0x0000e] in M:\BuildAgent\work\6c3239faf2b92630\src\NzbDrone.Console\ConsoleApp.cs:20

Press any key to exit...
```

I couldn't find where logs could be retrieved. I'd assume to get this figured out before looking at monit. (reference was more contextual.)

thx

**Comment by Dr-Bean** [comment]
Maybe a misunderstanding. You need to use `/var/packages/nzbdrone/scripts/start-stop-status start`. The start-stop-status script will handle the rest.
In this case, after a few seconds (the `sleep 2`), an mono process should be started. It won't show on the console: it's a daemon, not a console process. E.g. use `top` or `ps` to find out if the process is started. Another option is to use `/var/packages/nzbdrone/scripts/start-stop-status status`. Since Mono is quite a heavy process, it'll take a while before Sonarr is started completely and the web interface is available, so take that into account as well.

Don't try to run `/usr/local/mono/bin/mono --debug /usr/local/nzbdrone/share/NzbDrone/NzbDrone.exe`. First off: you're not switching to the correct user. If nothing else, you're essentially telling Sonarr to start from scratch: new database, new database location, and so on. 
Secondly, take another look at the command we use to start Sonarr: https://github.com/SynoCommunity/spksrc/blob/master/spk/sonarr/src/dsm-control.sh#L17. You're missing quite a bit of essential information there.

The Monit logs are available in `/usr/local/monit/var/monit.log`. That is also explained in the CONTRIBUTING documentation...

**Comment by fredguy** [comment]
Thanks, i was indeed doubting starting without the script was a good idea.
So back to the start script. 

this i what i get : 
Mon May 16 14:48:23 CEST 2016
> /var/packages/nzbdrone/scripts/start-stop-status status
Sonarr is not running
> /var/packages/nzbdrone/scripts/start-stop-status start
Starting Sonarr ...
> /var/packages/nzbdrone/scripts/start-stop-status status
Sonarr is not running
> date
Mon May 16 14:48:34 CEST 2016
> date
Mon May 16 14:51:19 CEST 2016
> /var/packages/nzbdrone/scripts/start-stop-status status
Sonarr is not running

It seems that the script is simply not achieving anything 

as the monit log i got the following : 

`[Debug] ProcessPro
[CEST May 11 18:20:52] error    : 'sonar' failed protocol test [HTTP] at [192.168.0.100]:8989 [TCP/IP] -- Connection refused
[CEST May 11 18:20:52] info     : 'sonar' trying to restart
[CEST May 11 18:20:52] info     : 'sonar' stop: /var/packages/nzbdrone/scripts/start-stop-status
[CEST May 11 18:20:52] info     : 'sonar' stopped
[CEST May 11 18:20:53] info     : 'sonar' start: /usr/local/mono/bin/mono
[CEST May 11 18:21:23] error    : 'sonar' failed to start (exit status -1) -- /usr/local/mono/bin/mono: Program timed out -- [Info] Bootstrap: Starting Sonarr - /volume1/@appstore/nzbdrone/share/NzbDrone/NzbDrone.exe - Version 2.0.0.3953
[Debug] ProcessProvider: Found 0 processes with the name: NzbDrone.Console`

**Comment by fredguy** [comment]
hmm better layout
```
[Debug] ProcessPro
[CEST May 11 18:20:52] error    : 'sonar' failed protocol test [HTTP] at [192.168.0.100]:8989 [TCP/IP] -- Connection refused
[CEST May 11 18:20:52] info     : 'sonar' trying to restart
[CEST May 11 18:20:52] info     : 'sonar' stop: /var/packages/nzbdrone/scripts/start-stop-status
[CEST May 11 18:20:52] info     : 'sonar' stopped
[CEST May 11 18:20:53] info     : 'sonar' start: /usr/local/mono/bin/mono
[CEST May 11 18:21:23] error    : 'sonar' failed to start (exit status -1) -- /usr/local/mono/bin/mono: Program timed out -- [Info] Bootstrap: Starting Sonarr - /volume1/@appstore/nzbdrone/share/NzbDrone/NzbDrone.exe - Version 2.0.0.3953
[Debug] ProcessProvider: Found 0 processes with the name: NzbDrone.Console
```

**Comment by Dr-Bean** [comment]
Ok. `/var/packages/nzbdrone/scripts/start-stop-status start` is exactly what Package Center runs in the background when you press `start` (on DSM5 anyway). If that does work via PC, but not via the console, I'm just guessing here, but this may be related to permissions that prevent you from starting it yourself (could be related to your attempts to start nzbdrone outside of the start-stop-status script, but no matter)

When the package is in the `stopped` state, check the following:
- Make sure there are no `mono` processes running (use `ps -w | grep mono` or equivalent);
- There should not be a PID file (when started, it'll be located at  `/usr/local/nzbdrone/var/.config/NzbDrone/nzbdrone.pid`)
- Please check and post the contents and permissions of `/usr/local/nzbdrone/var` (use `ls -al <dir>`). Do the same check for `/usr/local/nzbdrone/var/.config/NzbDrone`. I believe that by default, the files should be owned by `nzbdrone:users`.

**Comment by Dr-Bean** [comment]
@fredguy Any progress here?

**Comment by fredguy** [comment]
Hi There, Sorry for the long delay, i was travelling a bit lately. 
Somehow the PID file stays around for a while when the service stops. 

```
> ps -w | grep mono
16461 root      3772 S    grep mono

> pwd
/usr/local/nzbdrone/var
> ll -al
drwxr-sr-x    3 nzbdrone root          4096 Aug  2  2015 .
drwxrwxrwx    7 nzbdrone root          4096 Aug  2  2015 ..
drwxr-sr-x    4 nzbdrone root          4096 Apr 26  2014 .config
-rw-r--r--    1 root     root          2128 Aug  2  2015 install.log
-rw-r--r--    1 root     root             0 May 31  2014 nzbdrone.pid

> cd /usr/local/nzbdrone/var/.config/NzbDrone
> ll -al
drwxr-sr-x    6 nzbdrone root          4096 May 30 10:18 .
drwxr-sr-x    4 nzbdrone root          4096 Apr 26  2014 ..
drwxr-sr-x    4 nzbdrone root          4096 Oct  8  2014 Backups
drwxr-sr-x  123 nzbdrone root          4096 Apr 16 17:15 MediaCover
drwxr-sr-x    2 nzbdrone root          4096 May 12 03:28 UpdateLogs
-rw-r--r--    1 nzbdrone root           470 May 29 01:38 config.xml
drwxr-sr-x    2 nzbdrone root          4096 May 30 04:58 logs
-rw-r--r--    1 nzbdrone root      61462528 May 30 07:40 logs.db
-rw-r--r--    1 nzbdrone root        458752 May 30 10:18 logs.db-shm
-rw-r--r--    1 nzbdrone root      59872272 May 30 10:18 logs.db-wal
-rw-r--r--    1 nzbdrone root      18485248 May 30 10:18 nzbdrone.db
-rw-r--r--    1 nzbdrone root         32768 May 30 10:27 nzbdrone.db-shm
-rw-r--r--    1 nzbdrone root        100640 May 30 10:27 nzbdrone.db-wal
-rw-r--r--    1 nzbdrone root             5 May 29 01:38 nzbdrone.pi
```

At this moment the service is stopped, but i still have these 2 Pid files laying around.

**Comment by Remmie1973** [comment]
Hi, did you solve this problem?
I have the same
/var/packages/nzbdrone/scripts/start-stop-status start not working

**Comment by fredguy** [comment]
hi mate, the issue seemed to be a speed issue. it takes time for the syno to "process" more than expected. I added some cycles and time to the whole thing.

In the end this is my config and it does the job : 
Hope it helps. 

`set mailserver localhost port 25

 set mail-format {
     from: my@domain.com
     subject: $SERVICE $EVENT at $DATE
     message: Monit $ACTION $SERVICE at $DATE on $HOST: $DESCRIPTION.

 }
 set alert my@domain.com
 set alert //my api key//@api.pushover.net


check host sonar with address 192.168.0.100
every 2 cycles
start program = "/var/packages/nzbdrone/scripts/start-stop-status start" with timeout 300 seconds
stop program = "/var/packages/nzbdrone/scripts/start-stop-status stop"
if failed port 8989 protocol http then restart
if 3 restarts within 5 cycles then stop

#end
`

**Comment by Remmie1973** [comment]
Thanks for the quick response,
I will wait a bit longer :)
Dont think it will solve it for me, but thanks

**Comment by Safihre** [comment]
Closing this due to age.
Assuming newer `mono` packages solve underlying problems. Additionally a newer (DSM6 compatible) Sonarr/Radarr package is available in #3138. After the official release on SynoCommunity, the package will auto-update from the official release.
